### ISSUE TITLE
External forces in MPCD

### ISSUE THREAD

**Issue by mphoward** [glotzerlab/hoomd-blue#366]
## Description

MPCD doesn't use pair potentials for efficiency, but it can use external forces. MPCD should allow the user to put the MPCD particles in some sort of external force field, e.g., a constant body force to drive flow. 

## Proposed solution

The kernel that does the "streaming" (integration) of the MPCD particles is highly memory bound, and so it is not a good idea to split it into two Verlet-like kernels and load forces from memory. Rather, the external forces should just be computed on-the-fly in the kernel, since it is looking for more flops and usually external forces are cheap to evaluate anyway.

The most natural way to do this would be with some runtime solution, as I don't really want to have to compile the kernel for all possible streaming geometries (see #365) and external forces. Runtime compilation would be a possibility, but that seems like a bit of a hassle to get working. Using runtime polymorphism would be ideal, but CUDA does not allow passing objects with vtables between host and device. You can, however, have a polymorphic object in a kernel if it is constructed on the device.

I think that the most rapid development path would be to write a thin (templated) wrapper class that makes a mirror of a simple (likely const) object on both the host and the device. The device memory will be allocated on the host by the wrapper and a kernel will be called to construct the object in that memory using placement new. The wrapper can then return the appropriate pointer to host or device memory, which should be able to be passed to a kernel and dereferenced. The wrapper destructor will take care of freeing the memory. (It would be nice if there were a managed memory solution to do this more cleanly, but I haven't thought about / experimented with this.)

Once the wrapper is in place, it should be simple enough to implement an `mpcd.external` module with various forces and interface this with the MPCD integrator.

## Additional context

To be implemented after #365 is resolved.

## Developer

Assuming I can get the virtual wrapper working right, yes.

**Comment by mphoward** [comment]
I've implemented a preliminary version of the polymorphic wrapper on the `mpcd-force` branch. But, I ran into a really weird issue, and so I would appreciated some feedback from @joaander and @jglaser before I proceed down that route further.

I followed the directions I could find online as best I could related to polymorphism and CUDA, although these are sparse. The consensus seemed to be that you need to instantiate the object on the device, as CUDA forbids passing an object with a virtual table from the host to the device. I wrote some `device_new` and `device_delete` functions that I thought would do just that (cudaMalloc the memory, use placement new inside a kernel to stick the object into the pointer, etc.). Then, I've written a thin wrapper around these to manage the construction and deletion with unique pointers. I instantiated these in a `*.cu` file that went along with the header where I defined some virtual objects (e.g., `hoomd/mpcd/ExternalField.{h,cu}`).

Everything seemed to work OK with debug symbols on, but things went a little crazy when I switched to release mode. I started getting segfaults in any kernels using the polymorphic pointers. But, this only happened when `device_new` and `device_delete` were not explicitly instantiated in the same `*.cu` file as the kernel (and hence it didn't show up in my original unit tests for the wrapper). (This is with CUDA 8 and gcc 4.8.4 fwiw.) If I just stick those template instantiations in the file with the kernels, then everything runs just fine again.

I have no idea why this should be the case, and I'm not sure if it means I'm getting lucky (but actually doing something wrong), if I'm tricking the compiler into not doing some optimization it wants to do, or if there is some compiler quirk. Do you have any ideas?
